kimchi/src/circuits/cairo/runner/program.rs

```diff
@@ -0,0 +1,466 @@
+            } else if self.instr().opcode() == OPC_RET {
+                next_fp = Some(self.vars.dst.unwrap()); // ret sets fp to previous fp that was in [ap-2]
+            } else if self.instr().opcode() == OPC_AEQ {
+                // The following conditional is a fix that is not explained in the whitepaper
+                // The goal is to distinguish two types of ASSERT_EQUAL where one checks that
+                // dst = res , but in order for this to be true, one sometimes needs to write
+                // the res in mem(dst_addr) and sometimes write dst in mem(res_dir). The only
+                // case where res can be None is when res = op1 and thus res_dir = op1_addr
+                if self.vars.res.is_none() {
+                    self.mem.write(self.vars.op1_addr, self.vars.dst.unwrap()); // res = dst
+                } else {
+                    self.mem.write(self.vars.dst_addr, self.vars.res.unwrap()); // dst = res
+                }
+                next_fp = Some(self.curr.fp); // no modification on fp
+            } else {
+                unimplemented!();
+            }
+        } else {
+            unimplemented!(); // invalid instruction
+        }
+        (next_ap, next_fp)
+    }
+}
+
+/// This struct stores the needed information to run a program
+pub struct CairoProgram<'a, F: PrimeField> {
+    /// full execution memory
+    mem: &'a mut CairoMemory<F>,
+    /// initial computation registers regs
+    regs: CairoPointers<F>,
+}
+
+impl<'a, F: PrimeField> CairoProgram<'a, F> {
+    /// Creates a Cairo machine from the public input
+    pub fn new(mem: &mut CairoMemory<F>, pc: u64, ap: u64) -> CairoProgram<F> {
+        CairoProgram {
+            mem,
+            regs: CairoPointers::new(F::from(pc), F::from(ap), F::from(ap)),
+        }
+    }
+
+    /// This function simulates an execution of the Cairo program received as input.
+    /// It generates the full memory stack and the execution trace
+    pub fn execute(&mut self) {
+        // set finishing flag to false, as it just started
+        let mut end = false;
+        // saves local copy of the initial (claimed) registers of the program
+        let mut regs = self.regs;
+        // first timestamp
+        //let mut i = 0;
+        // keep executing steps until the end is reached
+        while !end {
+            // save allocation pointer before the execution
+            let curr_ap = regs.ap;
+            // create current step of computation
+            let mut step = CairoStep::new(self.mem, regs);
+            // execute current step and increase time counter
+            step.execute();
+            //i += 1;
+            match step.next {
+                None => end = true, // if find no next registers, end
+                _ => {
+                    // if there are next registers
+                    end = false;
+                    regs = step.next.unwrap();
+                    if curr_ap <= regs.pc {
```

Comment:
Since `ap` can be controlled by the program, it seems like this may not be the right termination condition (for example, you could write to some memory and then jump to it to execute it). Should we check if the memory at `pc` is `None` instead?

Reply:
Ok yeah, this is a tricky one and you got it :D Short answer is: yeah, it makes sense that the program finishes when the program counter points to a `None` address, meaning that there are no more instructions. But there are two problems with this: 
1. Instructions are given at compilation time. And in compilation time you still don't know how long the execution will take, so you cannot predict where there will be an address that is empty. 
2. Recall that the compiled program and all of the variables created in execution will share the same memory struct. If you knew in compilation time where there will be an empty cell, you would be leaking information about how long the computation takes. But this is input (witness) dependent, so you cannot really tell in compilation (and should not).
3. A correctness condition for the memory is that it has no holes. So you cannot really leave empty cells in the middle. In issue [294](https://github.com/o1-labs/proof-systems/issues/294) I give more intuition about the shape of the memory cells. 

So long answer would be: the last return of the program moves the program counter to `current+1`, and that address should be the first "empty" cell at the beginning of the execution (the first empty cell after compilation). And in reality, I have noticed that real Cairo executions tend to write in that cell the address where outputs are written. Thus, it's a cell that is written afterwards, some kind of retrospective. And the final `ap` points to that same address, where the outputs are written. So that's why I found that condition to be a candidate halting condition. 

Nonetheless, and as explained in the issue, I still need to carefully analyse their repo to make sure what's their mechanism to fill these cells and address halting.